### ISSUE TITLE
Unable to use Heldion 2.0 built-in health-check

### ISSUE THREAD

**Issue by bogedal** [oracle/helidon#2270]
Im trying to get the build-in health checks to work in my microprofile projectby including the following in my build.gradle file :

    implementation platform("io.helidon:helidon-dependencies:${project['io.helidon.version']}")
    implementation 'io.helidon.microprofile.bundles:helidon-microprofile-core'
    implementation 'io.helidon.microprofile.tracing:helidon-microprofile-tracing'
    implementation 'io.helidon.health:helidon-health-checks'
    implementation 'io.helidon.tracing:helidon-tracing-jaeger'

Note - (io.helidon.version = 2.0.1)

When I try and start the service using just **Server.create().start();** then I encounter two **WELD-001408: Unsatisfied dependencies**

* Helidon Version: **2.0.1**
* **Helidon MP**
* JDK version: **11**
* OS: **Windows**
* Docker version (if applicable):

----------

### Full startup stacktrace:
```java
Listening for transport dt_socket at address: 5002
aug. 18, 2020 3:20:47 PM io.helidon.common.LogConfig doConfigureLogging
INFO: Logging at initialization configured using defaults
2020-08-18 15:20:48,362 [INFO] [:No Correlation ID] org.jboss.weld.Version Thread[main]: WELD-000900: 3.1.3 (Final)
2020-08-18 15:20:49,463 [INFO] [:No Correlation ID] org.jboss.weld.Bootstrap Thread[main]: WELD-ENV-000020: Using jandex for bean discovery
2020-08-18 15:20:49,805 [INFO] [:No Correlation ID] org.jboss.weld.Bootstrap Thread[main]: WELD-000101: Transactional services not available. Injection of @Inject UserTransaction not available. Transactional observers will be invoked synchronously.
2020-08-18 15:20:50,039 [INFO] [:No Correlation ID] org.jboss.weld.Event Thread[main]: WELD-000411: Observer method [BackedAnnotatedMethod] public org.glassfish.jersey.ext.cdi1x.internal.ProcessAllAnnotatedTypes.processAnnotatedType(@Observes ProcessAnnotatedType<?>, BeanManager) receives events for all annotated types. Consider restricting events using @WithAnnotations or a generic type with bounds.
2020-08-18 15:20:50,804 [INFO] [:No Correlation ID] org.jboss.weld.Bootstrap Thread[main]: WELD-ENV-002001: Weld SE container fbd4abf2-6fd3-4a55-9fbf-f3ea48c4e382 shut down
Exception in thread "main" org.jboss.weld.exceptions.DeploymentException: Exception List with 2 exceptions:
Exception 0 :
org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type Runtime with qualifiers @Default
  at injection point [BackedAnnotatedParameter] Parameter 1 of [BackedAnnotatedConstructor] @Inject io.helidon.health.checks.HeapMemoryHealthCheck(Runtime, @ConfigProperty double)
  at io.helidon.health.checks.HeapMemoryHealthCheck.<init>(HeapMemoryHealthCheck.java:0)

        at org.jboss.weld.bootstrap.Validator.validateInjectionPointForDeploymentProblems(Validator.java:378)
        at org.jboss.weld.bootstrap.Validator.validateInjectionPoint(Validator.java:290)
        at org.jboss.weld.bootstrap.Validator.validateGeneralBean(Validator.java:143)
        at org.jboss.weld.bootstrap.Validator.validateRIBean(Validator.java:164)
        at org.jboss.weld.bootstrap.Validator.validateBean(Validator.java:526)
        at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:64)
        at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:62)
        at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:62)
        at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:55)
        at org.jboss.weld.executor.CommonForkJoinPoolExecutorServices.lambda$wrap$0(CommonForkJoinPoolExecutorServices.java:70)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1448)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Exception 1 :
org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type ThreadMXBean with qualifiers @Default
  at injection point [BackedAnnotatedParameter] Parameter 1 of [BackedAnnotatedConstructor] @Inject io.helidon.health.checks.DeadlockHealthCheck(ThreadMXBean)
  at io.helidon.health.checks.DeadlockHealthCheck.<init>(DeadlockHealthCheck.java:0)

        at org.jboss.weld.bootstrap.Validator.validateInjectionPointForDeploymentProblems(Validator.java:378)
        at org.jboss.weld.bootstrap.Validator.validateInjectionPoint(Validator.java:290)
        at org.jboss.weld.bootstrap.Validator.validateGeneralBean(Validator.java:143)
        at org.jboss.weld.bootstrap.Validator.validateRIBean(Validator.java:164)
        at org.jboss.weld.bootstrap.Validator.validateBean(Validator.java:526)
        at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:64)
        at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:62)
        at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:62)
        at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:55)
        at org.jboss.weld.executor.CommonForkJoinPoolExecutorServices.lambda$wrap$0(CommonForkJoinPoolExecutorServices.java:70)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1448)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)

        at org.jboss.weld.bootstrap.ConcurrentValidator.validateBeans(ConcurrentValidator.java:72)
        at org.jboss.weld.bootstrap.Validator.validateDeployment(Validator.java:487)
        at org.jboss.weld.bootstrap.WeldStartup.validateBeans(WeldStartup.java:496)
        at org.jboss.weld.bootstrap.WeldBootstrap.validateBeans(WeldBootstrap.java:93)
        at io.helidon.microprofile.cdi.HelidonContainerImpl.doStart(HelidonContainerImpl.java:293)
        at io.helidon.common.context.Contexts.runInContext(Contexts.java:118)
        at io.helidon.microprofile.cdi.HelidonContainerImpl.start(HelidonContainerImpl.java:245)
        at io.helidon.microprofile.cdi.Main.main(Main.java:80)
        Suppressed: org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type Runtime with qualifiers @Default
  at injection point [BackedAnnotatedParameter] Parameter 1 of [BackedAnnotatedConstructor] @Inject io.helidon.health.checks.HeapMemoryHealthCheck(Runtime, @ConfigProperty double)
  at io.helidon.health.checks.HeapMemoryHealthCheck.<init>(HeapMemoryHealthCheck.java:0)

                at org.jboss.weld.bootstrap.Validator.validateInjectionPointForDeploymentProblems(Validator.java:378)
                at org.jboss.weld.bootstrap.Validator.validateInjectionPoint(Validator.java:290)
                at org.jboss.weld.bootstrap.Validator.validateGeneralBean(Validator.java:143)
                at org.jboss.weld.bootstrap.Validator.validateRIBean(Validator.java:164)
                at org.jboss.weld.bootstrap.Validator.validateBean(Validator.java:526)
                at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:64)
                at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:62)
                at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:62)
                at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:55)
                at org.jboss.weld.executor.CommonForkJoinPoolExecutorServices.lambda$wrap$0(CommonForkJoinPoolExecutorServices.java:70)
                at java.base/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1448)
                at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
                at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
                at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
                at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
                at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
        Suppressed: org.jboss.weld.exceptions.DeploymentException: WELD-001408: Unsatisfied dependencies for type ThreadMXBean with qualifiers @Default
  at injection point [BackedAnnotatedParameter] Parameter 1 of [BackedAnnotatedConstructor] @Inject io.helidon.health.checks.DeadlockHealthCheck(ThreadMXBean)
  at io.helidon.health.checks.DeadlockHealthCheck.<init>(DeadlockHealthCheck.java:0)

                at org.jboss.weld.bootstrap.Validator.validateInjectionPointForDeploymentProblems(Validator.java:378)
                at org.jboss.weld.bootstrap.Validator.validateInjectionPoint(Validator.java:290)
                at org.jboss.weld.bootstrap.Validator.validateGeneralBean(Validator.java:143)
                at org.jboss.weld.bootstrap.Validator.validateRIBean(Validator.java:164)
                at org.jboss.weld.bootstrap.Validator.validateBean(Validator.java:526)
                at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:64)
                at org.jboss.weld.bootstrap.ConcurrentValidator$1.doWork(ConcurrentValidator.java:62)
                at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:62)
                at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:55)
                at org.jboss.weld.executor.CommonForkJoinPoolExecutorServices.lambda$wrap$0(CommonForkJoinPoolExecutorServices.java:70)
                at java.base/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1448)
                at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
                at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
                at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
                at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
                at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```
### Issue is easy to reproduce.
* Create a Helidon 2.0 project using JDK 11
* Include the implementations I showed in the top in the build.gradle file
* Start the service

Can you confirm this issue, or how to fix it please?

Kind regards
Bogedal

**Comment by barchetta** [comment]
**Short answer**: 

Replace `implementation 'io.helidon.health:helidon-health-checks'` with `implementation 'io.helidon.microprofile.health:helidon-microprofile-health`. `helidon-health-checks` is support for some built-in health checks -- not the MicroProfile support.

**Long answer**:

Helidon provides two bundles for MicroProfile support:

1. `helidon-microprofile-core`: this provides just basic MicroProfile support (JAX-RS, CDI, JSON, and MicroProfile config)
2. `helidon-microprofile`: this has everything

So an alternate solution would be do change your `helidon-microprofile-core` dependency to helidon-microprofile`.

If you use `-core` then you need to add the additional MicroProfile dependencies explicitly. So you would add dependencies like:

```
    implementation 'io.helidon.microprofile.health:helidon-microprofile-health'
    implementation 'io.helidon.microprofile.openapi:helidon-microprofile-openapi'
    implementation 'io.helidon.microprofile.metrics:helidon-microprofile-metrics'
```

And another tweak on `build.gradle`: change  `implementation platform` to ` implementation enforcedPlatform`. We have a bug in our example `build.gradle` files that's getting fixed in the next release. You can see the latest version here: https://github.com/oracle/helidon/blob/master/examples/quickstarts/helidon-quickstart-mp/build.gradle

**Comment by bogedal** [comment]
@barchetta Thanks a lot for your reply

I've tried to implement your suggestions but encountered a small issue

My build.gradle now looks like this :
```gradle
    implementation enforcedPlatform("io.helidon:helidon-dependencies:${project['io.helidon.version']}")
    implementation 'io.helidon.microprofile.bundles:helidon-microprofile-core'
    implementation 'io.helidon.microprofile.tracing:helidon-microprofile-tracing'
    implementation 'io.helidon.microprofile.health:helidon-microprofile-health'
    implementation 'io.helidon.microprofile.metrics:helidon-microprofile-metrics'
    implementation 'io.helidon.microprofile.openapi:helidon-microprofile-openapi'
    implementation 'io.helidon.tracing:helidon-tracing-jaeger'
```

However this seems to break CDI as I get a nullpointer exception in my application.java file.
I have a custom made Environment class that is injected in application class which now is null.

I breaktracked the issue to occur with I add the dependency to **'io.helidon.microprofile.openapi:helidon-microprofile-openapi'**

If I omit that then the service replies OK to both /metrics and /health also the service functions as expected.

**Comment by barchetta** [comment]
@bogedal I'm curious if changing the `helidon-microprofile-core` dependency to `helidon-microprofile` changes the behavior. This would narrow down if this is a dependency issue or something else. In any case, if you have a small reproducer that would help.

**Comment by bogedal** [comment]
@barchetta If I just use 
```gradle
    implementation enforcedPlatform("io.helidon:helidon-dependencies:${project['io.helidon.version']}")
    implementation 'io.helidon.microprofile.bundles:helidon-microprofile'
    implementation 'io.helidon.tracing:helidon-tracing-jaeger'
```
The issue is the same.
I'll try and create a small service to reproduce the issue. I'll get back to you again shortly

**Comment by bogedal** [comment]
@barchetta I've create a sample project which shows the same issue as above
[helidon-2-possible-cdi-issue.zip](https://github.com/oracle/helidon/files/5107071/helidon-2-possible-cdi-issue.zip)

Hope this can help track down the issue.

**Comment by barchetta** [comment]
Thanks for the reproducer, and sorry for not getting back to you sooner. The NPE problem looks like a bug in either our OpenAPI support, or our CDI support. I've created a new issue to track it: #2324 .

The issue is that our OpenAPI support performs application scanning during CDI initialization. Looks like it performs that scanning before Injection points have been resolved -- so if your `getClasses()` method uses an injection point it will be null.

Later during actual loading of your JAX-RS Application, your `getClasses()` method is called again -- but this time the injection points have been resolved so it works as expected.

The work-around is to code your `getClasses()` method to deal with this early scanning where injection points might be null. 

I'm going to close this issue since the original problem (using health check in a gradle project) has been resolved. See #2324 to track the null injected field  issue.
